File: packages/kits/ConvergenceKit/Tests/ConvergenceKitFederationTests/IntegrityHookTests.swift
// IntegrityHookTests.swift
//
// Four required tests for the post-apply integrity hook (R3, CVK-ICLOUD P2-M3).
//
// All tests use a two-engine, shared-relay rig identical to the pattern in
// FederationObserverOutboxTests — InMemoryStorage, FederationRelay, standard
// SyncManifest with the "items" table. The hook is installed on the RECEIVING
// engine's manifest; the sending engine uses a plain manifest without a hook.
//
// Behavioral contract verified:
//   R3-1  hook invoked once per pull batch with correct AppliedBatch contents
//   R3-2  hook writes carry origin == .local and flow into the outbox (ship)
//   R3-3  hook throw counts as ONE additional conflict; pull cycle does NOT abort
//   R3-4  empty-batch rule: hook NOT invoked when zero records are applied

import Testing
import Foundation
@testable import ConvergenceKitFederation
import ConvergenceKit
import SubstrateTypes
import PersistenceKit
import PersistenceKitInMemory

// MARK: - Thread-safe invocation tracker

/// Actor used to count hook calls and capture the last batch from the closure.
/// Storing AppliedBatch is safe: Storage is Sendable by protocol constraint and
/// the struct is @unchecked Sendable.
private actor HookTracker: Sendable {
    private(set) var invocationCount = 0
    private(set) var capturedAppliedByTable: [String: [UUID]] = [:]
    private(set) var capturedDeletedByTable: [String: [UUID]] = [:]

    func record(batch: AppliedBatch) {
        invocationCount += 1
        capturedAppliedByTable = batch.appliedByTable
        capturedDeletedByTable = batch.deletedByTable
    }
}

@Suite("Integrity hook — R3 post-apply behavioral contract")
struct IntegrityHookTests {

    // MARK: - Rig helpers

    static let schema = SchemaDeclaration(
        kitID: "HookTestKit",
        version: 1,
        tables: [
            TableDeclaration(
                name: "items",
                columns: [.uuid("id"), .text("note"), .bitmap("flags")],
                primaryKey: ["id"]
            )
        ]
    )

    func makeStorage() async throws -> any Storage {
        let s = InMemoryStorage(configuration: EstateConfiguration(
            estateID: UUID(), backend: .inMemory
        ))
        try await s.open(schema: IntegrityHookTests.schema)
        return s
    }

    func plainManifest() -> SyncManifest {
        SyncManifest(
            kitID: "HookTestKit",
            schemaVersion: 1,
            zoneIdentifier: "hook-test-zone",
            tables: [SyncedTable(name: "items", primaryKeyColumn: "id",
                                 conflictPolicy: .lastWriterWinsByHLC)]
        )
    }

    func manifestWithHook(_ hook: @Sendable @escaping (AppliedBatch) async throws -> Void) -> SyncManifest {
        var m = plainManifest()
        m.postApplyIntegrityHook = hook
        return m
    }

    /// Push until a non-zero pushed count or the 2 s deadline passes.
    /// The observer Task populates the outbox asynchronously; the retry loop
    /// mirrors the Rust test's bounded poll.
    func pushUntilNonzero(_ engine: FederationSyncEngine) async throws -> Int {
        let deadline = Date().addingTimeInterval(2.0)
        while true {
            let pushed = try await engine.push().pushed
            if pushed > 0 || Date() >= deadline { return pushed }
            try await Task.sleep(nanoseconds: 10_000_000)
        }
    }

    // MARK: - R3-1: hook invoked once per batch with correct AppliedBatch contents

    @Test("R3-1: hook invoked once per pull batch with correct applied/deleted keys")
    func hookInvokedOnceWithCorrectContents() async throws {
        let storageA = try await makeStorage()
        let storageB = try await makeStorage()

        let tracker = HookTracker()
        let relay = FederationRelay()
        let engineA = FederationSyncEngine(relay: relay)
        let engineB = FederationSyncEngine(relay: relay)
        try await engineA.enable(manifest: plainManifest(), storage: storageA)
        try await engineB.enable(
            manifest: manifestWithHook { batch in await tracker.record(batch: batch) },
            storage: storageB
        )
        try await engineA.pair(with: engineB, family: HyperplaneFamilySpec(seed: 0x1234_5678))
        defer { Task { try? await engineA.disable(); try? await engineB.disable() } }

        // Engine A inserts a row — the observer populates A's outbox.
        let insertedID = UUID()
        _ = try await storageA.rowStore.upsert(
            table: "items",
            values: ["id": .uuid(insertedID), "note": .text("hello"), "flags": .bitmap(0)],
            conflictColumns: ["id"]
        )

        // Engine A pushes; engine B pulls — hook fires exactly once.
        let pushed = try await pushUntilNonzero(engineA)
        #expect(pushed >= 1, "A must push at least one record")
        _ = try await engineB.pull()

        let callCount = await tracker.invocationCount
        #expect(callCount == 1, "hook must be invoked exactly once per pull batch")

        let appliedByTable = await tracker.capturedAppliedByTable
        let appliedIDs = appliedByTable["items"] ?? []
        #expect(appliedIDs.contains(insertedID),
                "applied batch must include the inserted row key under 'items'")

        let deletedByTable = await tracker.capturedDeletedByTable
        #expect(deletedByTable.isEmpty || (deletedByTable["items"] ?? []).isEmpty,
                "no deletes were sent, so deletedByTable must be empty")
    }

    // MARK: - R3-2: hook writes carry origin == .local and ship on next push

    /// Kong Q2 adjudication (hook-writes-must-ship): writes made through
    /// AppliedBatch.storage use the non-sync-tagged paths and carry origin
    /// == .local. The storage observer fires, populates the outbox, and a
    /// subsequent push delivers the repair record to the peer.
    @Test("R3-2: hook writes carry origin == .local and flow into the outbox")
    func hookWritesShipOnNextPush() async throws {
        let storageA = try await makeStorage()
        let storageB = try await makeStorage()

        // The repair row that the hook writes.
        let repairID = UUID()

        let relay = FederationRelay()
        let engineA = FederationSyncEngine(relay: relay)
        let engineB = FederationSyncEngine(relay: relay)
        try await engineA.enable(manifest: plainManifest(), storage: storageA)
        // B's hook: when a batch arrives, write a repair row to storage.
        // The repair write uses the non-sync-tagged upsert path, so origin == .local.
        try await engineB.enable(
            manifest: manifestWithHook { batch in
                _ = try await batch.storage.rowStore.upsert(
                    table: "items",
                    values: ["id": .uuid(repairID), "note": .text("repaired"), "flags": .bitmap(0xFF)],
                    conflictColumns: ["id"]
                )
            },
            storage: storageB
        )
        try await engineA.pair(with: engineB, family: HyperplaneFamilySpec(seed: 0xABCD_EF01))
        defer { Task { try? await engineA.disable(); try? await engineB.disable() } }

        // A inserts → B pulls → hook fires → hook writes repairID into storageB.
        let seedID = UUID()
        _ = try await storageA.rowStore.upsert(
            table: "items",
            values: ["id": .uuid(seedID), "note": .text("seed"), "flags": .bitmap(0)],
            conflictColumns: ["id"]
        )
        let pushed = try await pushUntilNonzero(engineA)
        #expect(pushed >= 1)
        _ = try await engineB.pull()

        // Verify the repair row exists in B's storage before any push.
        let repairRows = try await storageB.rowStore.query(
            table: "items",
            where: .eq(Column(table: "items", name: "id"), .uuid(repairID))
        )
        #expect(repairRows.count == 1, "hook must have written the repair row to storageB")

        // Give the storage observer Task time to enqueue the repair write into B's
        // outbox. The observer fires asynchronously; under parallel test load the
        // cooperative thread pool needs an explicit yield window before the push
        // attempt (mirrors the 100ms sleep in FederationStubTests push regression).
        try await Task.sleep(nanoseconds: 200_000_000)

        // Push B → A: the repair row must propagate (hook-writes-must-ship).
        let repairPushed = try await pushUntilNonzero(engineB)
        #expect(repairPushed >= 1, "repair row must flow into B's outbox and ship to A")
        _ = try await engineA.pull()

        let repairOnA = try await storageA.rowStore.query(
            table: "items",
            where: .eq(Column(table: "items", name: "id"), .uuid(repairID))
        )
        #expect(repairOnA.count == 1, "repair row must arrive at A via push/pull")
    }

    // MARK: - R3-3: hook throw counts as one conflict, pull does not abort

    @Test("R3-3: hook throw counts as one conflict and does not abort the pull cycle")
    func hookThrowCountsAsConflict() async throws {
        let storageA = try await makeStorage()
        let storageB = try await makeStorage()

        struct HookError: Error {}

        let relay = FederationRelay()
        let engineA = FederationSyncEngine(relay: relay)
        let engineB = FederationSyncEngine(relay: relay)
        try await engineA.enable(manifest: plainManifest(), storage: storageA)
        // B's hook always throws.
        try await engineB.enable(
            manifest: manifestWithHook { _ in throw HookError() },
            storage: storageB
        )
        try await engineA.pair(with: engineB, family: HyperplaneFamilySpec(seed: 0xDEAD_BEEF))
        defer { Task { try? await engineA.disable(); try? await engineB.disable() } }

        // A inserts → B pulls (hook throws).
        let rowID = UUID()
        _ = try await storageA.rowStore.upsert(
            table: "items",
            values: ["id": .uuid(rowID), "note": .text("trigger"), "flags": .bitmap(0)],
            conflictColumns: ["id"]
        )
        let pushed = try await pushUntilNonzero(engineA)
        #expect(pushed >= 1)
        let receipt = try await engineB.pull()

        // The record was applied (pulled == 1); hook throw added one conflict.
        #expect(receipt.pulled >= 1,
                "pull must have applied the record before the hook ran")
        #expect(receipt.conflicts == 1,
                "hook throw must count as exactly one conflict in the receipt")

        // Confirm the row reached B despite the hook failure.
        let rows = try await storageB.rowStore.query(
            table: "items",
            where: .eq(Column(table: "items", name: "id"), .uuid(rowID))
        )
        #expect(rows.count == 1, "record must be present even though the hook threw")
    }

    // MARK: - R3-4: empty-batch rule — hook NOT invoked when zero records applied

    @Test("R3-4: hook is not invoked when zero records were applied (empty-batch rule)")
    func hookNotInvokedForEmptyBatch() async throws {
        let storageB = try await makeStorage()

        let tracker = HookTracker()
        let engineB = FederationSyncEngine()
        try await engineB.enable(
            manifest: manifestWithHook { batch in await tracker.record(batch: batch) },
            storage: storageB
        )
        defer { Task { try? await engineB.disable() } }

        // Pull with nothing in the relay: zero records applied.
        _ = try await engineB.pull()

        let callCount = await tracker.invocationCount
        #expect(callCount == 0,
                "hook must NOT be invoked when the pull batch applied zero records")
    }
}

File: packages/kits/ConvergenceKit/rust/tests/integrity_hook_tests.rs
// integrity_hook_tests.rs
//
// Parity tests for the post-apply integrity hook (R3, CVK-ICLOUD P2-M3).
// Mirrors IntegrityHookTests.swift for the four required behavioral contracts.
//
// R3-1  hook invoked once per pull batch with correct AppliedBatch contents
// R3-2  hook writes flow into the observer outbox and ship on next push
// R3-3  hook throw (Err return) counts as ONE additional conflict; pull does not abort
// R3-4  empty-batch rule: hook NOT invoked when zero records are applied

use std::collections::BTreeMap;
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant};
use persistence_kit::{
    inmemory::InMemoryStorage, Column, ColumnDeclaration, ColumnType, SchemaDeclaration,
    Storage, StoragePredicate, TableDeclaration, TypedValue,
};
use convergence_kit::{
    AppliedBatch, ConflictPolicy, FederationRelay, FederationSyncEngine, HyperplaneFamilySpec,
    IntegrityHookFn, LocalIdentity, SyncDirection, SyncEngine, SyncError, SyncManifest, SyncedTable,
};
use uuid::Uuid;

// ----- helpers ---------------------------------------------------------------

fn make_storage() -> Arc<dyn Storage> {
    let storage = Arc::new(InMemoryStorage::with_estate(Uuid::new_v4()));
    let schema = SchemaDeclaration::new(
        "hook-test-kit",
        1,
        vec![TableDeclaration::new(
            "items",
            vec![
                ColumnDeclaration::new("id", ColumnType::Uuid),
                ColumnDeclaration::new("note", ColumnType::Text),
            ],
            vec!["id".to_string()],
        )],
    );
    storage.open(&schema).expect("open items schema");
    storage
}

fn plain_manifest() -> SyncManifest {
    SyncManifest::new(
        "hook-test-kit",
        1,
        "hook-test-zone",
        vec![SyncedTable::new("items", "id")
            .with_direction(SyncDirection::Bidirectional)
            .with_conflict_policy(ConflictPolicy::LastWriterWinsByHLC)],
    )
}

fn manifest_with_hook(hook: IntegrityHookFn) -> SyncManifest {
    plain_manifest().with_integrity_hook(hook)
}

/// Pair two engines symmetrically over a shared relay.
fn make_pair(
    storage_a: Arc<dyn Storage>,
    storage_b: Arc<dyn Storage>,
    manifest_b: SyncManifest,
) -> (FederationSyncEngine, FederationSyncEngine) {
    let relay = Arc::new(FederationRelay::new());
    let id_a = Arc::new(LocalIdentity::generate());
    let id_b = Arc::new(LocalIdentity::generate());
    let mut engine_a = FederationSyncEngine::new(id_a, relay.clone());
    let mut engine_b = FederationSyncEngine::new(id_b, relay.clone());
    engine_a.enable(plain_manifest(), storage_a).unwrap();
    engine_b.enable(manifest_b, storage_b).unwrap();

    // Symmetric pairing: both directions needed for bidirectional flow.
    let family = HyperplaneFamilySpec::new(0xBEEF);
    engine_a.pair(&engine_b, family).unwrap();
    engine_b.pair(&engine_a, family).unwrap();

    (engine_a, engine_b)
}

fn write_row(storage: &Arc<dyn Storage>, row_id: Uuid, note: &str) {
    let mut values: BTreeMap<String, TypedValue> = BTreeMap::new();
    values.insert("id".to_string(), TypedValue::Uuid(row_id));
    values.insert("note".to_string(), TypedValue::Text(note.to_string()));
    storage
        .row_store()
        .upsert("items", values, &["id".to_string()])
        .expect("upsert row");
}

fn row_count(storage: &Arc<dyn Storage>, row_id: Uuid) -> usize {
    let pred = StoragePredicate::Eq(Column::new("items", "id"), TypedValue::Uuid(row_id));
    storage.row_store().count("items", Some(&pred)).unwrap_or(0)
}

/// Retry push until nonzero pushed count or 2-second deadline.
/// Observer workers run on background threads; the first push may race ahead.
fn push_until_nonzero(engine: &mut FederationSyncEngine) -> usize {
    let deadline = Instant::now() + Duration::from_secs(2);
    loop {
        let pushed = engine.push().unwrap().pushed;
        if pushed > 0 || Instant::now() >= deadline {
            return pushed;
        }
        std::thread::sleep(Duration::from_millis(10));
    }
}

// ----- R3-1: hook invoked once with correct AppliedBatch contents -----------

#[test]
fn r3_1_hook_invoked_once_with_correct_contents() {
    let storage_a = make_storage();
    let storage_b = make_storage();

    // Tracker: (call_count, applied_keys_for_items)
    let tracker: Arc<Mutex<(usize, Vec<Uuid>)>> = Arc::new(Mutex::new((0, vec![])));
    let tracker_clone = Arc::clone(&tracker);

    let hook: IntegrityHookFn = Arc::new(move |batch: &AppliedBatch| {
        let mut t = tracker_clone.lock().unwrap();
        t.0 += 1;
        if let Some(keys) = batch.applied_by_table.get("items") {
            t.1.extend_from_slice(keys);
        }
        Ok(())
    });

    let (mut engine_a, mut engine_b) = make_pair(
        storage_a.clone(),
        storage_b.clone(),
        manifest_with_hook(hook),
    );

    let inserted_id = Uuid::new_v4();
    write_row(&storage_a, inserted_id, "hello");

    let pushed = push_until_nonzero(&mut engine_a);
    assert!(pushed >= 1, "A must push at least one record");

    let receipt = engine_b.pull().unwrap();
    assert!(receipt.pulled >= 1, "B must have applied the record");

    let t = tracker.lock().unwrap();
    assert_eq!(t.0, 1, "hook must be invoked exactly once per pull batch");
    assert!(
        t.1.contains(&inserted_id),
        "applied_by_table[items] must contain the inserted row key"
    );
}

// ----- R3-2: hook writes flow into outbox and ship on next push -------------

#[test]
fn r3_2_hook_writes_ship_on_next_push() {
    let storage_a = make_storage();
    let storage_b = make_storage();

    let repair_id = Uuid::new_v4();

    let hook: IntegrityHookFn = Arc::new(move |batch: &AppliedBatch| {
        // Write a repair row using the hook's storage handle.
        // This write runs after pulling.store(false), so it is NOT suppressed
        // by the echo-suppression guard — it flows into the outbox normally.
        let mut values: BTreeMap<String, TypedValue> = BTreeMap::new();
        values.insert("id".to_string(), TypedValue::Uuid(repair_id));
        values.insert("note".to_string(), TypedValue::Text("repaired".to_string()));
        batch
            .storage
            .row_store()
            .upsert("items", values, &["id".to_string()])
            .map_err(|e| SyncError::TransportFailure { detail: e.to_string() })?;
        Ok(())
    });

    let (mut engine_a, mut engine_b) = make_pair(
        storage_a.clone(),
        storage_b.clone(),
        manifest_with_hook(hook),
    );

    // A inserts → B pulls → hook fires → hook writes repairID to storageB.
    let seed_id = Uuid::new_v4();
    write_row(&storage_a, seed_id, "seed");

    let pushed = push_until_nonzero(&mut engine_a);
    assert!(pushed >= 1);

    let _receipt = engine_b.pull().unwrap();

    // Verify repairID is in storageB before any push.
    assert_eq!(row_count(&storage_b, repair_id), 1, "hook must have written repair row");

    // Give the observer worker thread time to enqueue the hook write.
    // The write happens synchronously during pull, but the observer reads
    // the channel on a 100 ms tick — allow up to 200 ms for pick-up.
    std::thread::sleep(Duration::from_millis(200));

    // Push B → A: the repair row must propagate (hook-writes-must-ship).
    let repair_pushed = push_until_nonzero(&mut engine_b);
    assert!(repair_pushed >= 1, "repair row must flow into B's outbox and ship to A");

    let _receipt_a = engine_a.pull().unwrap();
    assert_eq!(
        row_count(&storage_a, repair_id),
        1,
        "repair row must arrive at A via the push/pull cycle"
    );
}

// ----- R3-3: hook Err counts as one conflict; pull does not abort -----------

#[test]
fn r3_3_hook_err_counts_as_one_conflict_and_does_not_abort() {
    let storage_a = make_storage();
    let storage_b = make_storage();

    // Hook always fails.
    let hook: IntegrityHookFn = Arc::new(|_batch: &AppliedBatch| {
        Err(SyncError::TransportFailure { detail: "integrity check failed".to_string() })
    });

    let (mut engine_a, mut engine_b) = make_pair(
        storage_a.clone(),
        storage_b.clone(),
        manifest_with_hook(hook),
    );

    let row_id = Uuid::new_v4();
    write_row(&storage_a, row_id, "trigger");

    let pushed = push_until_nonzero(&mut engine_a);
    assert!(pushed >= 1);

    let receipt = engine_b.pull().unwrap();

    // Record was applied (pulled >= 1), hook added one conflict.
    assert!(receipt.pulled >= 1, "record must be applied before the hook runs");
    assert_eq!(receipt.conflicts, 1, "hook Err must count as exactly one conflict");

    // Row must be in storageB despite the hook failure.
    assert_eq!(row_count(&storage_b, row_id), 1, "record must persist even when hook fails");
}

// ----- R3-4: empty-batch rule — hook NOT invoked when zero records applied --

#[test]
fn r3_4_hook_not_invoked_for_empty_batch() {
    let storage_b = make_storage();

    let call_count: Arc<Mutex<usize>> = Arc::new(Mutex::new(0));
    let call_count_clone = Arc::clone(&call_count);

    let hook: IntegrityHookFn = Arc::new(move |_batch: &AppliedBatch| {
        *call_count_clone.lock().unwrap() += 1;
        Ok(())
    });

    // Enable B alone (no peer). Pull with nothing in the relay.
    let relay = Arc::new(FederationRelay::new());
    let id_b = Arc::new(LocalIdentity::generate());
    let mut engine_b = FederationSyncEngine::new(id_b, relay);
    engine_b.enable(manifest_with_hook(hook), storage_b).unwrap();

    let receipt = engine_b.pull().unwrap();
    assert_eq!(receipt.pulled, 0, "no records were available to pull");

    let count = *call_count.lock().unwrap();
    assert_eq!(count, 0, "hook must NOT be invoked when zero records were applied");
}
